vim-patch:9.1.{0258,0260,0280}

luukvbaal opened the pull request:
**vim-patch:9.1.0258: half-page scrolling broke backward compatibility**

Problem:  Support for 'smoothscroll' in (half-)page scrolling
          broke backward compatibility and can be made to work better.
          (after v9.1.215)
Solution: Restore the previous cursor and end-of-buffer behavior for
          half-page scrolling and improve 'smoothscroll' support.
          (Luuk van Baal)

https://github.com/vim/vim/commit/cb204e688e5c9d56a78b621ef27c35d91860cb09

**vim-patch:9.1.0260: Problems with "zb" and scrolling to new topline with 'smoothscroll'**

Problem: "zb" does not reveal filler lines at the start of a buffer.
          Scrolled cursor position with 'smoothscroll' is unpredictable,
          and may reset skipcol later if it is not visible (after v9.1.258)
Solution: Replace confusing for loop that reaches final control value too
          early with while loop. Set "w_curswant" accordingly so cursor
          will be placed in visible part of topline.
          (Luuk van Baal)

https://github.com/vim/vim/commit/bd28cae1f1c21c0e3743e3427c98bbd848fad237

**vim-patch:9.1.0280: several issues with 'smoothscroll' support**

Problem:  Logic to make sure cursor is in visible part of the screen after
          scrolling the text with 'smoothscroll' is scattered, asymmetric
          and contains bugs.
Solution: Adjust and create helper function for 'smoothscroll' cursor logic.
          (Luuk van Baal)

https://github.com/vim/vim/commit/9148ba8a46baa3934c44164989cdcdec5d01d9e3

Fix #28106
Fix #28115
Fix #28211

luukvbaal committed:
vim-patch:9.1.0258: half-page scrolling broke backward compatibility

Problem:  Support for 'smoothscroll' in (half-)page scrolling
          broke backward compatibility and can be made to work better.
          (after v9.1.215)
Solution: Restore the previous cursor and end-of-buffer behavior for
          half-page scrolling and improve 'smoothscroll' support.
          (Luuk van Baal)

https://github.com/vim/vim/commit/cb204e688e5c9d56a78b621ef27c35d91860cb09

zeertzjq commented on src/nvim/move.c:
```diff
@@ -2341,61 +2352,117 @@ static int get_scroll_overlap(Direction dir)
-/// Move screen "count" pages up ("dir" is BACKWARD) or down ("dir" is FORWARD)
-/// and update the screen.
+/// Scroll "count" lines with 'smoothscroll' in direction "dir". Adjust "count"
+/// when scrolling more than "count" and return TRUE when scrolling happened.
+static int scroll_with_sms(Direction dir, int *count)
```
```suggestion
/// when scrolling more than "count" and return true when scrolling happened.
static bool scroll_with_sms(Direction dir, int *count)
```

luukvbaal committed:
vim-patch:9.1.0260: Problems with "zb" and scrolling to new topline with 'smoothscroll'

Problem: "zb" does not reveal filler lines at the start of a buffer.
          Scrolled cursor position with 'smoothscroll' is unpredictable,
          and may reset skipcol later if it is not visible (after v9.1.258)
Solution: Replace confusing for loop that reaches final control value too
          early with while loop. Set "w_curswant" accordingly so cursor
          will be placed in visible part of topline.
          (Luuk van Baal)

https://github.com/vim/vim/commit/bd28cae1f1c21c0e3743e3427c98bbd848fad237

BlankTiger commented:
I just compiled neovim straight from `luukvbaal:vim-9.1.0258` and while `<C-u>zz` and `<C-d>zz` work great, the same issue still persists with `<C-f>zz` and `<C-b>zz`. Essentially half-page scrolling seems to be fixed, but full-page scrolling still seems broken.

zeertzjq commented:
What is your expected behavior? Please create a separate issue.

BlankTiger commented:
It already has an issue: https://github.com/neovim/neovim/issues/28106. Basically it's broken scrolling but for full-page scroll mappings + zz (which is in scope of the mentioned issue because it doesn't specify that only half-page scrolling + zz is broken).

Sorry if it's the wrong place for this comment, I can add it in the mentioned issue if that would be more appropriate.

zeertzjq commented:
Please answer the question: What behavior do you expect?

BlankTiger commented:
Expected behavior would be the same as described in https://github.com/neovim/neovim/issues/28106: For the mappings to scroll and center the cursor.

Problematic mappings are the the same as in the mentioned issue with the difference being that `<C-d>` (half-page scroll downwards) becomes `<C-f>` (full-page scroll downwards), and `<C-u>` (half-page scroll upwards) becomes `<C-b>` (full-page scroll upwards):

```lua
vim.keymap.set('n', '<C-f>', '<C-f>zz', { desc = 'Scroll downwards' })
vim.keymap.set('n', '<C-b>', '<C-b>zz', { desc = 'Scroll upwards' })
```

zeertzjq commented:
That's a reasonable feature request, but I don't think they are not supposed to behave like that, so there is no bug.

zeertzjq commented:
I tested on older versions and it seems that they are not supposed to behave so, so there is no bug.

BlankTiger commented:
Weird, works for me on stable nvim 0.9.5 appimage with `--clean` flag and only those mapping set.

luukvbaal commented:
If you see difference in behavior please provide clear reproduction steps (buffer text, window size, options, keys pressed).

luukvbaal committed:
vim-patch:9.1.0280: several issues with 'smoothscroll' support

Problem:  Logic to make sure cursor is in visible part of the screen after
          scrolling the text with 'smoothscroll' is scattered, asymmetric
          and contains bugs.
Solution: Adjust and create helper function for 'smoothscroll' cursor logic.
          (Luuk van Baal)

https://github.com/vim/vim/commit/9148ba8a46baa3934c44164989cdcdec5d01d9e3

luukvbaal commented on src/nvim/plines.c:
```diff
@@ -885,6 +885,9 @@ int plines_m_win(win_T *wp, linenr_T first, linenr_T last, bool limit_winheight)
   }
+  if (first == wp->w_buffer->b_ml.ml_line_count + 1) {
+    count += win_get_fill(wp, first);
+  }
```
Over at vim/vim it was [suggested](https://github.com/vim/vim/pull/14410#discussion_r1554805999) to not add this to `plines_win()`, and decided to instead add it to `plines_m_win()`. Going by [this](https://github.com/vim/vim/pull/14410#discussion_r1554853125) logic, it might make more sense for us to add it to `plines_win_full()`? Or do we still have concerns of it causing problems there?

zeertzjq replied:
But that'll lead to incorrect `w_cline_height`.

zeertzjq commented on src/nvim/move.c:
```diff
@@ -1185,6 +1189,122 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+/// the screen with 'smoothscroll'.
+static void cursor_correct_sms(win_T *wp)
+{
+  if (!curwin->w_p_sms || !curwin->w_p_wrap
+      || curwin->w_cursor.lnum != curwin->w_topline) {
```
```suggestion
  if (!wp->w_p_sms || !wp->w_p_wrap
      || wp->w_cursor.lnum != wp->w_topline) {
```

zeertzjq commented on src/nvim/move.c:
```diff
@@ -2341,61 +2404,120 @@ static int get_scroll_overlap(Direction dir)
-    // Temporarily set 'smoothscroll' so that scrolling count lines
-    // does not skip over parts of the buffer with wrapped lines.
-    curwin->w_p_sms = true;
     if (count > 0) {
```
```suggestion
    // Scroll the window and determine number of lines to move the cursor.
    if (count > 0) {
```

zeertzjq commented on src/nvim/move.c:
```diff
@@ -1185,6 +1189,121 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+  int so_cols = so == 0 ? 0 : width1 + (so - 1) * width2;
+  int space_cols = (wp->w_height_inner - 1) * width2;
+  int size = so == 0 ? 0 : win_linetabsize(wp, wp->w_topline,
+                                           ml_get(wp->w_topline), (colnr_T)MAXCOL);
```
```suggestion
                                           ml_get_buf(wp->w_buffer, wp->w_topline), (colnr_T)MAXCOL);
```

zeertzjq replied:
There is also another one in `scrolldown()`, maybe correct that one as well.

luukvbaal replied:
Oh yeah, spotted that one as well.

zeertzjq commented on src/nvim/change.c:
```diff
@@ -343,8 +343,7 @@ static void changed_common(buf_T *buf, linenr_T lnum, colnr_T col, linenr_T lnum
               || (wp->w_topline >= lnum
                   && wp->w_topline < lnume
                   && win_linetabsize(wp, wp->w_topline, ml_get(wp->w_topline), MAXCOL)
```
```suggestion
                  && win_linetabsize(wp, wp->w_topline, ml_get_buf(wp->w_buffer, wp->w_topline), MAXCOL)
```

luukvbaal replied:
Lol, should have done this tomorrow not just before bed. There was another offender in move.c from the recent decurwinification.
